Add moar (#10)
andrsd/moose: modules/thermal_hydraulics/joss_paper.md

# Documentation

> THM is extensively documented following the same object-oriented paradigm as the code base. Each

Reviewer: ```suggestion
THM's documentation is extensive and follows the same structure as the code base. Each
```

object is accompanied by a documentation page that describes it, including its equations or a
figure as relevant, that lists its parameters along with metadata about the parameters, that lists

> all the input files in the repository that use this object, and finally all the classes deriving from this object.

Reviewer: Sentence is a little hard to read - make it a list instead

> object is accompanied by a documentation page that describes it, including its equations or a
> figure as relevant, that lists its parameters along with metadata about the parameters, that lists
> all the input files in the repository that use this object, and finally all the classes deriving from this object.

Author: ```suggestion
object is accompanied by a documentation page that:

- describes it, including its equations or a figure as relevant
- lists its parameters along with metadata about the parameters
- lists all the input files in the repository that use this object
- lists all the classes deriving from this object.
```

> Major groups of objects, usually derived from a single base class, are documented through the syntax documentation.
> Syntax documentation describes how to add a class of object to a simulation, and describes the common characteristics
> of this class of objects. Components and Closures are examples of a unique syntax in THM that also corresponds to
> base classes of groups of THM objects.

Reviewer: ```suggestion
Major groups of objects, usually derived from a single base class, are documented through the syntax documentation,
which describes how these objects are instantiated and used in a simulation.
For example, `Components` and `Closures` are examples of syntax unique to THM that also correspond to
base classes of groups of THM objects.
```

This documentation page is hosted on the module website [@thm_website]. The website also notably hosts the software
quality assurance (SQA) records, such as the testing requirement matrix or the failure analysis reports for example.

> The interested reader is referred to INL's SQA plan [@sqa] for more information.

Author: ```suggestion
The interested reader is referred to the MOOSE SQA plan [@sqa] for more information.
```

Reviewer: I think it would be fine to say "the MOOSE SQA plan"

# Testing

The module relies on CIVET [@slaughter2021continuous] for continuous integration.
Every pull request to the module runs the entire test suite for MOOSE, including the tests for THM.
The test suite is comprehensive and aims to cover every feature available in the module. It
consists of unit and regression tests, described below. The entire test suite is run with a wide variety of
configurations, from compiling with the oldest and newest supported compilers, to running with several
shared memory threads and distributed memory processes, on a variety of operating systems. The test suite
is also run encapsulated in Valgrind to detect memory leaks. Every pull request to the module
must include tests to cover any new features. This is ensured by the reviewer. The review process is detailed

> in INL's SQA plan [@sqa].

Reviewer: ```suggestion
in MOOSE's SQA plan [@sqa].
```

Unit tests in THM are targeted at specific routines that can be accessed by creating the relevant object with
example parameters. For example, creating a `Flux` object, we can check that the formulation of the numerical
flux, whether with a centered scheme or with HLLC, are both consistent and symmetric.

Regression tests are essentially created for each object to ensure that their output does not vary on a
relevant test case. For example, each function or each auxiliary kernel is tested on a simple Cartesian mesh to make sure
the field it produces is consistently the same. Components are often tested in the minimal configuration sufficient

> to satisfy the test requirement, to prove conservation of mass and momentum for example for a flow channel.

Reviewer: ```suggestion
to satisfy the test requirement, for example, to prove conservation of mass and energy on a flow channel.
```

> Finally, the test suite includes numerous Jacobian tests to prove the correctness of the implementation of the

Reviewer: I think we probably don't need to talk of Jacobian testing since it falls under "regression testing". However, that does remind me that it might be worth mentioning AD somewhere, but it's not too important I think.

> Finally, the test suite includes numerous Jacobian tests to prove the correctness of the implementation of the
> components with regards to computing the Jacobian of the nonlinear system. This implementation has historically
> been difficult, but exact numerical Jacobians are now computed in THM using automatic differentiation [@lindsay2021automatic].
> 

Author: ```suggestion

```

the usability of MOOSE's `Controls` system, allowing control units to be
chained together.

The development of the module is driven by user application needs.
Future work to THM may include
improvement of existing components, as well as additional components related to
single-phase flow and heat conduction. Depending on future needs, additional
flow models may be added as well. Further abstraction of the way the discretization
of an equation is created on a component is underway, and should allow for the definition
of a general multiphysics component, able to instantiate any equation discretized in MOOSE.

External contributions to the module are encouraged and

> supported to provide compliance with INL's SQA standard.

Reviewer: ```suggestion
External contributions to the module are encouraged, and
support will be provided to comply with MOOSE's SQA standard.
```

> supported to provide compliance with INL's SQA standard.

Author: ```suggestion
supported to provide compliance with MOOSE's SQA standard.
```

the usability of MOOSE's `Controls` system, allowing control units to be
chained together.

The development of the module is driven by user application needs.
Future work to THM may include
improvement of existing components, as well as additional components related to
single-phase flow and heat conduction. Depending on future needs, additional
flow models may be added as well. Further abstraction of the way the discretization
of an equation is created on a component is underway, and should allow for the definition
of a general multiphysics component, able to instantiate any equation discretized in MOOSE.

External contributions to the module are encouraged and

> supported to provide compliance with INL's SQA standard.

Reviewer: Not sure what you mean to say with the SQA part. Do you mean that welcoming the contributions is to provide compliance? Or do you mean that we welcome contributions, but they must comply with SQA?

Reviewer: @GiudGiud did you see this one?

Author: no

Author: so I mean that because we have such a stringent standard, the typical contributor will be helped so that theiur contribution makes it in depsite the standard
